File: app/src/main/java/com/mobarak/todo/ui/tasks/TasksFragment.java
package com.mobarak.todo.ui.tasks;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;

import com.mobarak.todo.R;
import com.mobarak.todo.data.AppRepositoryImpl;
import com.mobarak.todo.databinding.TasksFragBinding;
import com.mobarak.todo.ui.base.ViewModelFactory;
import com.mobarak.todo.utility.ViewUtil;

public class TasksFragment extends Fragment {

    private TasksViewModel viewModel;
    private TasksFragBinding viewDataBinding;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.tasks_frag, container, false);
        viewDataBinding = TasksFragBinding.bind(root);
        setHasOptionsMenu(true);
        return root;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = new ViewModelProvider(this,
                new ViewModelFactory(getContext(), AppRepositoryImpl.getInstance(), this)
        ).get(TasksViewModel.class);
        viewDataBinding.setLifecycleOwner(this);
        viewDataBinding.setViewmodel(viewModel);
        ViewUtil.setupRefreshLayout(getActivity(), viewDataBinding.refreshLayout, null);
        setupFab();
        setupListAdapter();
        setupSnackbar();

    }

    private void setupSnackbar() {
        if (getActivity() != null)
            viewModel.getSnackbarText().observe(getActivity(), message -> {
                ViewUtil.showSnackbar(viewDataBinding.getRoot(), message);
            });
    }

    @Override
    public void onStart() {
        super.onStart();
        viewModel.setFiltering(FilterType.ALL_TASKS);
        viewModel.loadTasks();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.tasks_fragment_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_clear:
//                viewModel.clearCompletedTasks();
                return true;

            case R.id.menu_filter:
                showFilteringPopUpMenu();
                return true;
            case R.id.menu_refresh:
                viewModel.loadTasks();
                return true;
            default:
                return false;
        }
    }


    private void showFilteringPopUpMenu() {
        View view = getActivity().findViewById(R.id.menu_filter);
        PopupMenu popup = new PopupMenu(getActivity(), view);
        popup.getMenuInflater().inflate(R.menu.filter_tasks, popup.getMenu());
        popup.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.active:
                    viewModel.filterItems(FilterType.ACTIVE_TASKS);
                    return true;
                case R.id.completed:
                    viewModel.filterItems(FilterType.COMPLETED_TASKS);
                    return true;
                case R.id.all:
                    viewModel.filterItems(FilterType.ALL_TASKS);
                    return true;
            }
            return false;
        });
        popup.show();
    }

    private void setupFab() {
        viewDataBinding.getRoot().findViewById(R.id.add_task_fab).setOnClickListener(view -> {
            NavDirections action = TasksFragmentDirections
                    .actionTasksFragmentToAddEditTaskFragment(-1, getActivity().getString(R.string.add_task));
            Navigation.findNavController(view).navigate(action);
        });
    }


    private void setupListAdapter() {
        if (viewModel != null) {
            TasksAdapter listAdapter = new TasksAdapter(viewModel);
            viewDataBinding.tasksList.setAdapter(listAdapter);
        } else {
            Log.i("", "ViewModel not initialized when attempting to set up adapter.");
        }
    }

}

File: app/src/main/java/com/mobarak/todo/ui/tasks/TasksFragment.kt
package com.mobarak.todo.ui.tasks

import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.PopupMenu
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.mobarak.todo.R
import com.mobarak.todo.databinding.TasksFragBinding
import com.mobarak.todo.utility.EventObserver
import com.mobarak.todo.utility.getViewModelFactory
import com.mobarak.todo.utility.setupRefreshLayout
import com.mobarak.todo.utility.setupSnackbar
import timber.log.Timber

class TasksFragment : Fragment() {
    private val viewModel by viewModels<TasksViewModel> { getViewModelFactory() }
    private lateinit var viewDataBinding: TasksFragBinding
    private val args: TasksFragmentArgs by navArgs()


    private lateinit var listAdapter: TasksAdapter
    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?, savedInstanceState: Bundle?): View {
        viewDataBinding = TasksFragBinding.inflate(inflater, container, false).apply {
            viewmodel = viewModel
        }
        setHasOptionsMenu(true)
        return viewDataBinding.root
    }

    override fun onOptionsItemSelected(item: MenuItem) =
            when (item.itemId) {
                R.id.menu_clear -> {
                    viewModel.clearCompletedTasks()
                    true
                }
                R.id.menu_filter -> {
                    showFilteringPopUpMenu()
                    true
                }
                R.id.menu_refresh -> {
                    viewModel.loadTasks(true)
                    true
                }
                else -> false
            }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.tasks_fragment_menu, menu)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // Set the lifecycle owner to the lifecycle of the view
        viewDataBinding.lifecycleOwner = this.viewLifecycleOwner
        setupSnackbar()
        setupListAdapter()
        setupRefreshLayout(viewDataBinding.refreshLayout, viewDataBinding.tasksList)
        setupNavigation()
//        setupFab()
    }

    private fun setupNavigation() {
        viewModel.openTaskEvent.observe(this.viewLifecycleOwner, EventObserver {
            openTaskDetails(it)
        })
        viewModel.newTaskEvent.observe(this.viewLifecycleOwner, EventObserver {
            navigateToAddNewTask()
        })
    }

    private fun setupSnackbar() {
        view?.setupSnackbar(this, viewModel.snackbarText, Snackbar.LENGTH_SHORT)
        arguments?.let {
            viewModel.showEditResultMessage(args.userMessage)
        }
    }

    private fun showFilteringPopUpMenu() {
        val view = activity?.findViewById<View>(R.id.menu_filter) ?: return
        PopupMenu(requireContext(), view).run {
            menuInflater.inflate(R.menu.filter_tasks, menu)

            setOnMenuItemClickListener {
                viewModel.setFiltering(
                        when (it.itemId) {
                            R.id.active -> FilterType.ACTIVE_TASKS
                            R.id.completed -> FilterType.COMPLETED_TASKS
                            else -> FilterType.ALL_TASKS
                        }
                )
                true
            }
            show()
        }
    }

    private fun setupFab() {
        view?.findViewById<FloatingActionButton>(R.id.add_task_fab)?.let {
            it.setOnClickListener {
                navigateToAddNewTask()
            }
        }
    }

    private fun navigateToAddNewTask() {
        val action = TasksFragmentDirections
                .actionTasksFragmentToAddEditTaskFragment(
                        0,
                        resources.getString(R.string.add_task)
                )
        findNavController().navigate(action)
    }

    private fun openTaskDetails(taskId: Long) {
        val action = TasksFragmentDirections.actionTasksFragmentToTaskDetailFragment(taskId)
        findNavController().navigate(action)
    }

    private fun setupListAdapter() {
        val viewModel = viewDataBinding.viewmodel
        if (viewModel != null) {
            listAdapter = TasksAdapter(viewModel)
            viewDataBinding.tasksList.adapter = listAdapter
        } else {
            Timber.w("ViewModel not initialized when attempting to set up adapter.")
        }
    }
}